RVC.cpp

```diff
@@ -0,0 +1,119 @@
+#include "RVC.h"
+#include "ByteConverter.h"
+
+RVC::RVC(const uint32_t id, const bool criticality, const uint32_t readInterval, Adafruit_BNO08x_RVC* rvc)
+{
+    this->id = id;
+    this->criticality = criticality;
+    this->readInterval = readInterval;
+    this->rvc = rvc;
+    heading = new BNO08x_RVC_Data;
+}
+
+RVC::~RVC() { delete heading; }
+
+bool RVC::ready()
+{
+    if (millis() - lastRead >= readInterval)
+    {
+        return rvc->read(heading);
+    }
+    return false;
+}
+
+void RVC::pack(uint8_t* buf, const float value, const uint8_t id)
+{
+    buf[0] = id;
+    ByteConverter::floatToBytes(value, &buf[1]);
+}
+
+float RVC::unpack(const uint8_t* buf)
+{
+    uint8_t bytes[sizeof(float)];
+    for (size_t i = 0; i < sizeof(float); i++)
+    {
+        bytes[i] = buf[i + 1]; // All bytes after first (prepended id)
+    }
+    return ByteConverter::floatFromBytes(bytes);
+}
+
+SensorData RVC::read()
+{
+    lastRead = millis();
+    SensorData sensorData = SensorData(id, 6);
+
+    constexpr size_t len = sizeof(float) + 1;
+    uint8_t bufferIndex = 0; // used for prependId and message index
+
+    uint8_t x_accelBuf[len];
+    pack(x_accelBuf, heading->x_accel, bufferIndex);
+    sensorData.setMsg(x_accelBuf, len, bufferIndex);
+    bufferIndex++;
+
+    uint8_t y_accelBuf[len];
+    pack(y_accelBuf, heading->y_accel, bufferIndex);
+    sensorData.setMsg(y_accelBuf, len, bufferIndex);
+    bufferIndex++;
+
+    uint8_t z_accelBuf[len];
+    pack(z_accelBuf, heading->z_accel, bufferIndex);
+    sensorData.setMsg(z_accelBuf, len, bufferIndex);
+    bufferIndex++;
+
+    uint8_t yawBuf[len];
+    pack(yawBuf, heading->yaw, bufferIndex);
+    sensorData.setMsg(yawBuf, len, bufferIndex);
+    bufferIndex++;
+
+    uint8_t pitchBuf[len];
+    pack(pitchBuf, heading->pitch, bufferIndex);
+    sensorData.setMsg(pitchBuf, len, bufferIndex);
+    bufferIndex++;
+
+    uint8_t rollBuf[len];
+    pack(rollBuf, heading->roll, bufferIndex);
+    sensorData.setMsg(rollBuf, len, bufferIndex);
```

Comment:
idk how to make this a comment instead of a "review comment" but just a note for both of us. We have never sent back to back messages like this before so we just gotta remember to check on the receiving end its handled right and none are missed.

Reply:
Agreed, but I don't think it will matter too much. The way I have it structured in my DataCollector that is working is like this:
```
#include "DataCollector.h"

void DataCollector::checkSensors() const
{
    for (size_t i = 0; i < numSensors; i++)
    {
        if (Sensor* sensor = sensors[i]; sensor->ready()) { readSensor(sensor); }
    }
}

void DataCollector::readSensor(Sensor* sensor) const
{
    const SensorData data = sensor->read();
    const size_t msgCount = data.getMsgCount();
    const CAN_message_t* msgs = data.getMsgs();
    for (size_t i = 0; i < msgCount; i++)
    {
        sensor->debugPrint(msgs[i]);
        can.write(msgs[i]);
    }
}
```

As for the receiving end, the RVC has the same ID for all of the messages, meaning that all of those messages come from the same sensor. The first buffer has the "subSensorId" the way I think of it, where 0 is X accel, 1 is Y accel, etc. We can implement specifics if we need to.